#198 Add next and previous links to a project page

ilya-konanykhin opened the pull request:
* Added the prev/next links
* Added tests covering this functionality
* Added translations for all locales (en/ru by hand, others via Google Translate)

![2018-08-03 14 28 29](https://user-images.githubusercontent.com/8665049/43632957-879ceec4-9729-11e8-918d-439c3b5d82b6.png)

Fixes #198

ilya-konanykhin commented:
There is one commit message in Russian unfortunately (one word meaning "refactoring"). I can squash all commits into one, give it a single English name and re-send a pull request. Please let me know if you need this.

jcn commented on app/helpers/project_nav_helper.rb:
```diff
@@ -0,0 +1,32 @@
+  private
+
+  def chapter_projects_keys(project)
+    project.chapter.projects.order('funded_on DESC').pluck(:id)
```
Also, this should be `project.chapter.winning_projects` - since this is on a public project page, it should only show "funded" projects (i.e. ones that have a `funded_on` date set). The tests should validate this as well.

ilya-konanykhin replied:
> Is there a reason this isn't memoized?

In general, I try to follow a functional approach, so that methods do not share state. Esp. helpers, since they are stateless modules per se, and their state object (`self`) might be not what we expect.

The proper way IMO is to use either cache or ADT. Both seemed like an overkill to me, the queries are simple and the DB is small. (Also, I've checked other helpers, they seem to not care about DB queries much as well.)

If you think caching this computation as an instance variable is fine, I can do that.

ilya-konanykhin replied:
> Also, this should be project.chapter.winning_projects

Thanks, this wasn't clear from the code :+1:

jcn commented on app/views/projects/public_show.html.erb:
```diff
@@ -3,6 +3,13 @@
+  <% prev_project, next_project = prev_project_id(@project), next_project_id(@project) %>
+  <% if prev_project || next_project %>
+    <div class="prev-next-links">
+      <%= link_to t('.prev-project'), project_path(prev_project), :class => 'prev-project' if prev_project %>
```
We actually need the full project object here since the canonical URL includes a project slug (see https://github.com/awesomefoundation/awesomebits/blob/master/app/models/project.rb#L103-L105)

ilya-konanykhin replied:
Right now this URL gets redirected to a proper canonical (guess by the projects controller).

I wanted to keep it that way for simplicity.

But ok, let's do a full URL from the beginning.

jcn commented on app/assets/stylesheets/_projects-show.scss:
```diff
@@ -11,6 +11,19 @@ header {
+    .next-project{
+      float: right;
+    }
+  }
```
This isn't working in Firefox (latest, OS X). Also, it doesn't appear to be picking up the link styles from the rest of the site (links are rendering blue underlined instead of pink and not underlined).

![screen shot 2018-08-06 at 11 05 36 am](https://user-images.githubusercontent.com/2909/43724808-eb8409e0-9968-11e8-8c4c-cf1d223ec73f.png)

ilya-konanykhin replied:
Firefox: ok, checking.

Links: you mean, the Next/Prev Project links should be pink, like here for example? https://www.awesomefoundation.org/en/about_us

I'll do, but you sure this is a good idea? We seem to have navigation or utility links gray/black all over the website, and only inline links are pink (inside a page content).

ilya-konanykhin commented:
Thanks for the feedback @jcn! Replying inline.

ilya-konanykhin commented on app/helpers/project_nav_helper.rb:
```diff
@@ -0,0 +1,32 @@
+module ProjectNavHelper
+
+  def next_project_id project
```
My bad! Got used to this style. Changing.

ilya-konanykhin replied:
(probably we should add `.rubocop.yml` and set up Rubocop, to ease following our code style)

ilya-konanykhin commented on spec/helpers/project_nav_helper.rb:
```diff
@@ -0,0 +1,73 @@
+    end
+  end
+
+  context 'private methods' do
```
Can't say I agree 100%, but ok, let's remove.

ilya-konanykhin commented:
Think I addressed all the issues @jcn, please take a look.

![2018-08-09 15 25 37](https://user-images.githubusercontent.com/8665049/43890406-7bf0381a-9be8-11e8-8b47-53ab16548511.png)

spacemunkay commented:
Wooot!! 🙌  Thanks @jcn and @ilya-konanykhin!
